Build system changes for BSD

robxnano opened the pull request:
These changes are mostly to fix building on the different BSD variants, plus some changes to the build scripts to remove the bash requirement. As BSD and macOS don't have bash installed by default, it means one less program to install. I also added some colors to the configure script, which aren't strictly necessary but make the output a little easier to read. This change can be removed if you'd prefer to keep it as it is.

**Tested on:**

- [x] Windows 10+  (via MinGW)
- [x] macOS 10.13+
- [x] Ubuntu Linux
- [x] FreeBSD
- [x] NetBSD
- [x] OpenBSD

robxnano committed:
configure.py: Use colors for terminal output

- The control characters for the colored text will only be used if
the output is being sent to a terminal
- Error messages have been consolidated into a single function

robxnano committed:
configure.py: Show passthrough arguments

- Don't turn arguments of the form --arg=value into exports

robxnano committed:
Make scripts POSIX compatible

Rewrote repo-info.sh, tag-release.sh and build-presets.sh to remove
the need for bash when compiling.

robxnano committed:
NetBSD: Build fixes

- Update python versions
- Add libjpeg-turbo to modules to avoid conflict
- Fix failure to make metainfo file

robxnano committed:
configure.py: Simplify BSD versions

robxnano committed:
Simplify modules in main.defs

robxnano committed:
Remove support for BeOS

robxnano committed:
Drop pthread conditionals and unused Cygwin code

robxnano committed:
configure.py: Simplify GTK flags

robxnano committed:
Remove unneeded BSD conditionals

robxnano committed:
OpenBSD: Fix warnings

robxnano committed:
Add shebangs to translation scripts

robxnano committed:
scripts: Remove Python 2 support

robxnano committed:
gtk: Fix LTO warnings

robxnano committed:
configure: Pass launcher return code to shell

Needed to detect failures in CI.

robxnano committed:
contrib/x264: Fix X264_BUILD number

There needs to be only one space before the build number.
Otherwise the shared library gets built without a version number.

galad87 commented:
I think @brad0 worked a bit on BSD compatibility too, they might want to take a look too.

galad87 commented:
Everything looks fine on macOS too.

sr55 commented:
LGTM as well. Thanks

bradleysepos commented on make/configure.py:
```diff
@@ -1409,19 +1409,19 @@ def createCLI( cross = None ):
-    h = IfHost( 'disable GTK GUI', '*-*-linux*', '*-*-*bsd*', none=argparse.SUPPRESS ).value
-    grp.add_argument( '--disable-gtk', default=False, action='store_true', help=h )
+    # GTK GUI is disabled by default on macOS and Windows, enabled otherwise
+    require_gtk = not (host_tuple.match('*-*-mingw*', '*-*-cygwin*', '*-*-darwin*'))
```
I'm not sure I like inverting the logic here. Would rather allowlist BSD and Linux.

robxnano replied:
I understand where you're coming from. I'll change it.

bradleysepos commented on make/configure.py:
```diff
@@ -1748,9 +1748,6 @@ class Tools:
     options.enable_ffmpeg_aac = IfHost(options.enable_ffmpeg_aac, '*-*-darwin*',
                                        none=True).value
-    # Allow GTK mingw only on mingw
-    options.enable_gtk_mingw  = IfHost(options.enable_gtk_mingw, '*-*-mingw*',
```
I take it we can no longer cross compile GTK? We actually did use this in the past, but perhaps it's no longer relevant.

robxnano replied:
It can still be done, but now it requires the `--enable-gtk` flag. Last I checked, it did build with some manual intervention, provided the necessary MinGW packages were installed.

bradleysepos replied:
Unless something is broken/unfixable here, we shouldn't rip out support for cross-compiling the GTK GUI using mingw-w64. It's not popular but people use it, and in the past we've used it for development purposes.

robxnano replied:
But the support is still in there? The only difference is that the flag has changed (and it applies to any operating system that doesn't enable GTK by default).

bradleysepos replied:
I think I see what is tripping me up here. What seems logical to me would be to:

- Rename `requires_gtk` to something like `gtk_default_enabled`, default `True` for BSD/Linux which sets `enable_gtk` to `True`, allows disabling via `disable_gtk` (already done aside from renaming)
- Add `gtk_supported` internal var, `True` for BSD/Linux/mingw, allow enabling/disabling only on these systems
- Update help output at bottom to display `note_unsupported` where `gtk_supported` is `False` (I'd also move this item to the end of the list, after libdovi)

Hope that makes some sense.

bradleysepos replied:
Something in the section `# Only allow these features on supported platforms` would probably make the most sense for gating this.

robxnano replied:
@bradleysepos
> Add `gtk_supported` internal var, `True` for BSD/Linux/mingw, allow enabling/disabling only on these systems

I'm not sure if this is necessary. The GTK-specific code should be portable enough to compile on any operating system. If someone wants to build HandBrake on an unsupported OS, it'll be easier if they can just enable the UI with a flag, rather than have to edit the script to add the OS manually.

bradleysepos replied:
That's fine

bradleysepos replied:
I guess my concern is if we're going to add a line to the bottom output showing GTK GUI is enabled/disabled, would hate for someone to think it's supported on their obscure platform because they see other instances of "not supported on target platform". So it's a consistency thing, perhaps. But not a hill I'm going to die on, either.

robxnano replied:
Maybe the message is a little misleading? IMO, "not supported on target platform" sounds like a feature that is available but not officially tested or supported by the developers. If it was something like "not available on target platform" instead, it would more accurately reflect that the feature cannot be enabled on that platform.

bradleysepos replied:
GTK GUI on systems other than BSD/Linux/mingw is "not supported" by that definition, so my previous suggestion still makes sense (to me at least). To date we have used the term to mean don't try it unless you are an engineer and want to tinker under the hood, so I guess I'm less enthused about changing the language here.

In the past, we've tried to encourage and allow only known good configurations, mostly because we're a small team and when things don't work, we get more support requests than we would like. Perhaps not ideal, but practical at times, and generally keeps everyday people from doing silly things.

robxnano replied:
In the end I thought it would be best to add the code to disable on other platforms for now, just to be on the safe side. I added `_supported` variables for each of the platform-specific features in order to simplify the script by not repeating the same definitions in three different places, which also makes it easier for anyone interested to enable support for another platform.

bradleysepos replied:
Seems reasonable, thanks.

bradleysepos commented on libhb/handbrake/ports.h:
```diff
@@ -152,10 +152,7 @@ void hb_get_user_config_filename( char name[1024], char *fmt, ... );
  ***********************************************************************/
 typedef struct hb_thread_s hb_thread_t;
 
-#if defined( SYS_BEOS )
```
Goodbye, old friend.

bradleysepos commented on configure:
```diff
@@ -13,23 +13,19 @@ inpath()
+for p in python3 python3.8 python3.7 python3.6 python3.5 python3.4 python3.3 python3.2 python3.1 python2 python2.7 python
+do
+    if ( inpath $p ); then
+        pp="$p"
```
Looks like this commit reversed some changes from the previous.

robxnano replied:
Sorry, but I'm not sure which changes you're referring to. The only changes I can see for `configure` are the one that removes the bash check, and the one that updates the python versions and simplifies the script to match `make/python_launcher`.

bradleysepos replied:
Looks like this was an indentation change that threw me off.

bradleysepos reviewed (changes requested):
Overall looks great, just a few comments. Thanks for doing this.

brad0 commented:
Although most looks Ok, a lot has little or nothing directly to do with *BSD. Like the changes for BeOS, POSIX threads, Python, POSIX compliancy for some scripts, etc. Typically I would split those up into smaller PRs.

brad0 commented on gtk/src/Makefile.am:
```diff
@@ -179,7 +179,7 @@ widget.deps: makedeps.py
 
 fr.handbrake.ghb.metainfo.xml: fr.handbrake.ghb.metainfo.xml.in
-	$(AM_V_GEN) $(MSGFMT) --xml --template $< -d $(top_srcdir)/po -o $@
+	$(AM_V_GEN) $(MSGFMT) --xml --template $< -d $(top_srcdir)/po -o $@ || cp $< $@
```
This looks odd.

robxnano replied:
This is a workaround for NetBSD, which has an old version of gettext which doesn't support xml. If `msgfmt` fails, it just copies the file without adding the localized strings. It was the best solution I could come up with.

bradleysepos replied:
Please add an inline code comment explaining why `cp` is used here.

robxnano replied:
Done.

robxnano committed:
Enable GTK by default on Linux and BSD only

It can be enabled on other operating systems with --enable-gtk

robxnano committed:
gtk: Add comment for NetBSD msgfmt workaround

robxnano committed:
configure: Allow GTK only on supported platforms

- Supported on Linux, FreeBSD, OpenBSD, NetBSD, and MinGW
- Simplified support checks in configure.py

bradleysepos commented:
Looks good now. Approved.

> Although most looks Ok, a lot has little or nothing directly to do with *BSD. Like the changes for BeOS, POSIX threads, Python, POSIX compliancy for some scripts, etc. Typically I would split those up into smaller PRs.

I have to agree with @brad0 here (Brads united).

This is a fairly large set of changes that I would prefer to be split into multiple PRs, but I also do not wish to stifle positive development this instance. Let's merge this and aim for smaller chunks in the future. Thanks @robxnano for the good work!

brad0 commented:
@robxnano Thank you for the work.
